prepress.py

```diff
@@ -255,17 +255,98 @@ def replace_dashes(article: Article) -> Article:
         text_tag.replace_with(new_tag)
     return article
 
+# Quotes only really start with varieties of left brackets
+LEFT_PUNCTUATION = { '(', '{', '<' }
+# But they can be ended with any of the standard sentence enders
+RIGHT_PUNCTUATION = { ')', '}', '>', '.', ',', '?', '!' }
+LEFT_DOUBLE_QUOTE = '“'
+RIGHT_DOUBLE_QUOTE = '”'
+STRAIGHT_DOUBLE_QUOTE = '"'
+
+def replace_double_quote(before: str, after: str):
+    # double quotes can occur at the beginning/end of strings
+    if before is None:
+        return LEFT_DOUBLE_QUOTE
+    if after is None:
+        return RIGHT_DOUBLE_QUOTE
+
+    # double quotes can occur before/after spaces
+    if before.isspace():
+        return LEFT_DOUBLE_QUOTE
+    # ditto for the right
+    if after.isspace():
+        return RIGHT_DOUBLE_QUOTE
+
+    # double quotes can occur before/after punctuation
+    if before in LEFT_PUNCTUATION:
+        return LEFT_DOUBLE_QUOTE
+    if after in RIGHT_PUNCTUATION:
+        return RIGHT_DOUBLE_QUOTE
+
+
+    return STRAIGHT_DOUBLE_QUOTE
+
+LEFT_SINGLE_QUOTE = '‘'
+RIGHT_SINGLE_QUOTE = '’'
+STRAIGHT_SINGLE_QUOTE = '\''
+
+def replace_single_quote(before: str, after: str):
+    # the rules for single quotes are the same as double quotes
+    double_quote = replace_double_quote(before, after)
+    if double_quote == RIGHT_DOUBLE_QUOTE:
+        return RIGHT_SINGLE_QUOTE
+    elif double_quote == LEFT_DOUBLE_QUOTE:
+        return LEFT_SINGLE_QUOTE
+    elif double_quote == STRAIGHT_DOUBLE_QUOTE:
+        return STRAIGHT_SINGLE_QUOTE
+
+
+def replace_smart_quotes(s: str):
+    # create an array so we can modify this string
+    char_array = list(s)
+    
+    for idx, char in enumerate(char_array):
+        before = None if idx == 0 else char_array[idx - 1]
+        after = None if idx == len(char_array) - 1 else char_array[idx + 1]
+        if char == '"':
+            char_array[idx] = replace_double_quote(before, after)
+        if char == '\'':
+            char_array[idx] = replace_single_quote(before, after)
+
+    return ''.join(char_array)
+    
 def add_smart_quotes(article: Article) -> Article:
-    """Replaces regular quotes with smart quotes. This function assumes quotes are not nested
-    and will simply convert pairs of quotes into left and right quotes.
-    """
-    text_tag: bs4.NavigableString
-    for text_tag in article.content.find_all(text=True):
-        if keep_verbatim(text_tag): continue
+    """Replaces regular quotes with smart quotes. Works on double and single quotes."""
+    text_tags: List[bs4.NavigableString] = list(article.content.find_all(text=True))
+    # some hackery here: breaks between text tags might lead to invalid quotes
+    # example: "|<em>text</em>|" will make the first quote a right quote, since
+    # it's at the end of its text tag.
+    # To avoid this, we glue the first character in the following tag
+    # and the last character in the previous tag to the current tag.
```

Comment:
Do we handle the case where only one quote may be encased in the text tag, e.g. `"<em>text"</em>`?

Reply:
So, the `<em>` tags don't actually get included in the text, so it should fallback to normal behavior, depending on what comes after that fragment.

My testing confirms this.